cci_cameroon/analysis/text_similarity/similarity_search_sample.py

```diff
@@ -0,0 +1,286 @@
+# %%
+sentences[:10]
+
+# %%
+# preprocessing the sentences fetched by removing duplicates and NaN
+sentences = [word for word in list(set(sentences)) if type(word) is str]
+
+# %%
+# initialize sentence transformer model - using BERT. Sentence embedding takes time
+model = SentenceTransformer(
+    "bert-base-nli-mean-tokens"
+)  # loading simple pre-trained transformer model for learning purposes
+# create sentence embeddings
+sentence_embeddings = model.encode(sentences)
+sentence_embeddings.shape
+
+# %%
+dimension = sentence_embeddings.shape[1]
+
+# %%
+# initialize the index with the dimension of the dataset
+index = faiss.IndexFlatL2(dimension)
+
+# %%
+sentence_embeddings.shape
+
+# %%
+index.is_trained
+
+# %%
+# we load the embeddings to the index
+index.add(sentence_embeddings)
+index.ntotal
+
+# %%
+sentences[:5]
+
+
+# %%
+def readInput():
+    txt = input("Enter text and hit enter : ")  # read user's input
+    return str(txt)
+
+
+# %%
+user_input = readInput()
+
+# %%
+k = 2  # we wish to return the index of the two closest strings to our query
+xq = model.encode([user_input])
+
+# %%
+# we now perform a search for similarity
+# %time
+distance, position = index.search(xq, k)  # search
+print(position)
+print(distance)
+
+# %%
+print("Similarity:", util.dot_score(xq, sentence_embeddings)[0][3465])
+
+# %%
+print("Similarity:", util.dot_score(xq, sentence_embeddings)[0][4118])
+
+# %%
+sentences[3465]
+
+# %%
+sentences[4118]
+
+# %% [markdown]
+# # Output using the Flat index with L2
+# * Algorithm runs for 5µs and returns top 2 nearest neighbors to the input text.
+# * First text has a similarity index of 57.92
+# * Second output has similarity index of 4.78
+# * Clearly, the sentence with a higher similarity is closer to the input query
+
+# %% [markdown]
+# # Drawback of  IndexFlatL2
+# It performs and exhaustive search, comparing the query string to all the vectors before returning the closest match.
+# The computation time quickly grows with the size of the dataset.
+# * We can partition the index to improve on the search time, although this would provide an approximate match as opposed
+# to the exhaustive search.
+
+# %%
+num_cells = 50  # Number of cells the index should be divided into
+quantizer = faiss.IndexFlatL2(dimension)
+index = faiss.IndexIVFFlat(quantizer, dimension, num_cells)
+
+# %%
+index.is_trained
+
+# %%
+index.train(sentence_embeddings)
+index.is_trained  # check if index is now trained
+
+# %%
+index.add(sentence_embeddings)
+index.ntotal  # number of embeddings indexed
+
+# %%
+# we now perform a search for similarity using the input text read earlier on.
+# %time
+distance, position = index.search(xq, k)  # search
+print(position)
+print(distance)
+
+# %%
+print("Similarity:", util.dot_score(xq[0], sentence_embeddings)[0][3465])
+
+# %%
+print("Similarity:", util.dot_score(xq[0], sentence_embeddings)[0][4118])
+
+# %% [markdown]
+# # By compressing the vectors, we further improve the computation time
```

Comment:
I like the description text throughout the document but the header sizes are a bit inconsistent which makes it a little hard to read (the one above is like title size but seems to be more a smaller sub section of the notebook).

Try and treat it like a word document where you have the title size at the top, then under that different subsections broken up using subtitles and anything nested under that smaller text or just plain text.